### ISSUE TITLE
LSP: rename results in "attempt to get length of a nil value" - likely when rename results in a new file

### ISSUE THREAD

**Issue by mfussenegger** [neovim/neovim#11790]
- `nvim --version`: v0.5.0-343-ge956ea767 / v0.5.0-338-g242af8420-dirty (second one has a locally modified runtime)
- `vim -u DEFAULTS` (version: ) behaves differently? Haven't tried, probably not relevant.
- Operating system/version: Linux 5.4.15-arch1-1 
- Terminal name/version: termite v15
- `$TERM`: xterm-256color

I haven't yet had time to take a closer look at it this, but opening it up early to see if anybody has an idea what's going on. 

When I try to rename a class (java) using `vim.lsp.buf.rename` I get an error raised at

```
  if set_eol and #lines[#lines] ~= 0 then
    table.insert(lines, '')
  end
``` 

within `M.apply_text_edits`  of `runtime/lua/vim/lsp/util.lua`.  `lines` is empty.

The same also occurs if I use [the code actions](https://github.com/neovim/neovim/pull/11607) and try to generate a new class(/file) using a code action.

I'll try to fill in some more details once I've some more time, and might also provide a PR with a fix if it is nothing too complex.

**Comment by Jesse-Bakker** [comment]
The neovim implementation of the protocol currently does not support file rename or file creation actions (see https://microsoft.github.io/language-server-protocol/specifications/specification-current/#workspaceEdit), only changes to existing documents (`TextDocumentEdit`s). You should have an error log saying `Unsupported change: ` with the change object.

I suspect you are using eclipse.jdt.ls for java, which uses a custom `WorkspaceEdit`. I guess they send a `java.apply.workspaceEdit` command when renaming a class to also rename the file, which, again, is not supported in neovim as of yet.

The LSP spec says that when certain client capabilities are not specified by the client when registering with the server, they should be interpreted as not being supported by the client. The server assumes all capabilities defined in the 2.x versions of the spec to be supported by the client, as registering capabilities was not available then. Neovim does not send any extra capabilities in the `workspace` section, which means only `TextEdit` edits are supported. This means that the server should not send those edits.

I have made some assumptions about your configuration and possible errors here, so feel free to correct me (and send along your logs so we can figure out the actual problem ;)), but I hope this helps.

**Comment by mfussenegger** [comment]
> I suspect you are using eclipse.jdt.ls for java, which uses a custom WorkspaceEdit

Yes you're right. In this case I used eclipse.jdt.ls. I tried to reproduce the rename case I had to verify if it is related to a custom WorkspaceEdit, but I couldn't reproduce it on a smaller project, so I switched to the codeAction case.

Due to your pointers I had a look again, and I don't think it is due to the lack of file rename or file creation handling. 


I extended the codeAction callback (that is being added in the other PR) to handle the jdtls `java.apply.workEdit`:

```
  if action.command == 'java.apply.workspaceEdit' then
    for _, argument in ipairs(action.arguments) do
      util.apply_workspace_edit(argument)
    end
```

Then on a file like this:

```
public class A {
    void foo() {
        B.bar(); // bar doesn't exist
    }
}
```

I use a code action to generate `bar`. What happens is that the action.arguments contains a ``TextEdit``. The call to `util.apply_workspace_edit(argument)` leads all the way to https://github.com/neovim/neovim/blob/master/runtime/lua/vim/lsp/util.lua#L91 which returns a empty table because the buffer for the file is not loaded. (I didn't open up `B` before triggering the code action).

If I make sure to open B first, the code action works.

If I add something like

```
  if not api.nvim_buf_is_loaded(bufnr) then
      api.nvim_command("buf " .. bufnr)
  end
```
above 
https://github.com/neovim/neovim/blob/master/runtime/lua/vim/lsp/util.lua#L91

the code action works as well. (I'm aware the code is bogus, I'm not very familiar with the API yet.)


After adding this code even code actions that create files work. E.g. having a 

```
public class A {
    void foo() {
        var c = new C(); // C doesn't exist yet
    }
}
```

And invoke a code action to generate `C` works succesfully. (Although there is then another error when trying to save a file, which also happens when I create a new file without using any code actions. I can create a separate issue for that.)
